### DISCUSSION TITLE
How to export typings from a node package

### DISCUSSION THREAD

**Discussion by verheyenkoen** [egoist/tsup#935] [General]
I'm converting a node API wrapper package (not published on NPM, I just install it using the GitHub path but that doesn't matter I guess) for [Aleph X-services](https://developers.exlibrisgroup.com/aleph/apis/Aleph-X-Services/) to Typescript. I wrote a file `aleph.d.ts` with typings for what the API returns (it responds XML and [xml2js](https://www.npmjs.com/package/xml2js) converts it to JS objects, which I need types for).

The typings file looks like this:

```ts
export declare namespace Aleph {
  type BorrowerInfoResponse {
    // ...
  }

  // ...
}
```

Whenever I build that with tsup (with the `--dst` flag), tsup removes the `export` keyword in the "dist/index.d.ts" file. The rest of the types is output as is to my dist folder, but not the `export` keyword, which makes it impossible to import these types from consumer projects. When I import the functions from same package in a Typescript file of a consumer project, types are inferred correctly (I can click to the definition), but in one case it is not, and I have to cast it manually which cannot be done if the Aleph namespace is not exported correctly.

I found, I can just rename the file `aleph.d.ts` to `aleph.ts` and add `export * from './aleph'` to my `index.ts` file and everything works but that doesn't feel right as it contains only typings for the API response structure.

Am I doing something wrong or is this a bug in tsup?

**Comment by baga89** [comment]
I'm facing a similar issue. I'm bundling a third-party package that contains a file with several enums, each prefixed with the export keyword. However, after the bundling process is complete, all the enums are missing the export keyword.

**Comment by benjick** [comment]
Any solution for this yet?

My file looks like this:

```ts
import createClient, { ClientOptions } from "openapi-fetch";
import type { paths as Radarr } from "./generated";

export function createRadarrClient(clientOptions?: ClientOptions) {
  return createClient<Radarr>(clientOptions);
}
```

But the `index.d.ts` just says `interface paths {` after build

**Comment by prisis** [comment]
You need to export the types in you index.ts file, without a export the tool dont know what should be exported

**Reply by benjick** [reply]
Wow, sorry about that 🤦 thank you!

**Reply by IsaiahByDayah** [reply]
So this works, but I'm still a bit confused as to _why_? I would imagine I should be able to export types from outside of just `index.ts`, but that doesn't seem to be the case 🤔

```ts
// a.ts
export interface INested {
  foo: boolean
}

=====================
// index.ts
export * as a from "./a.ts
```

I would expect I could then do the following but it doesn't seem to see that I exported `INested` from within `a.ts`:

```ts
import { type a } from "my-module"

const doThing = (args: a.INested) => { ... }
```

**Reply by IsaiahByDayah** [reply]
@prisis Any chance you could help explain why this doesn't work?

**Reply by prisis** [reply]
Hey :)

tsup works different then tsc (from typescript), tsup need to know witch files you like to exports, because its use the rollup plugin (rollup-plugin-dts) or "@microsoft/api-extractor" (if you choose experimental dts) to generate the types for you, out of the defined files that you provided. 

Just as example without the exports https://github.com/visulima/visulima/blob/main/packages/fs/src/index.ts in this file, tsup wouldnt generated the types

**Reply by IsaiahByDayah** [reply]
Hmmm, ok. Interesting! While a bit strange/less than ideal to me, since I can just re-export types from my index file it’s not a showstopper. I guess I just expected it to behave differently. Might dig into more later and see if there’s something I can config to work how I expected. Either way, thanks for the help!
